src/tokenpool-gnu-make-posix.cc

```diff
@@ -0,0 +1,368 @@
+  // add N tokens to pipe
+  const char token = '+'; // see make/posixos.c
+  while (parallelism--) {
+    if (write(wfd, &token, 1) < 1) {
+      close(wfd);
+      close(rfd);
+      return false;
+    }
+  }
+
+  // initialize file descriptors for this instance
+  rfd_ = rfd;
+  wfd_ = wfd;
+
+  return true;
+}
+
+int GNUmakeTokenPoolPosix::dup_rfd_ = -1;
+
+void GNUmakeTokenPoolPosix::CloseDupRfd(int signum) {
+  close(dup_rfd_);
+  dup_rfd_ = -1;
+}
+
+bool GNUmakeTokenPoolPosix::SetAlarmHandler() {
+  struct sigaction act;
+  memset(&act, 0, sizeof(act));
+  act.sa_handler = CloseDupRfd;
+  if (sigaction(SIGALRM, &act, &old_act_) < 0) {
+    perror("sigaction:");
+    return false;
+  }
+  restore_ = true;
+  return true;
+}
+
+bool GNUmakeTokenPoolPosix::ParseAuth(const char* jobserver) {
+  // check for jobserver-fifo style
+  const char* fifo;
+  if (((fifo = strstr(jobserver, "=fifo:")) != NULL) &&
+      CheckFifo(fifo + 6))
+    return SetAlarmHandler();
+
+  // check for legacy simple pipe style
+  int rfd = -1;
+  int wfd = -1;
+  if ((sscanf(jobserver, "%*[^=]=%d,%d", &rfd, &wfd) == 2) &&
+      CheckFd(rfd) &&
+      CheckFd(wfd) &&
+      SetAlarmHandler()) {
+    rfd_ = rfd;
+    wfd_ = wfd;
+    return true;
+  }
+
+  // some jobserver style we don't support
+  return false;
+}
+
+bool GNUmakeTokenPoolPosix::CreatePool(int parallelism,
+                                       const char* style,
+                                       std::string* auth) {
+  if (style == NULL || strcmp(style, "fifo") == 0)
+    return CreateFifo(parallelism, auth);
+
+  if (strcmp(style, "pipe") == 0)
+    return CreatePipe(parallelism, auth);
+
+  Fatal("unsupported tokenpool style '%s'", style);
+}
+
+bool GNUmakeTokenPoolPosix::AcquireToken() {
+  // Please read
+  //
+  //   http://make.mad-scientist.net/papers/jobserver-implementation/
+  //
+  // for the reasoning behind the following code.
+  //
+  // Try to read one character from the pipe. Returns true on success.
+  //
+  // First check if read() would succeed without blocking.
+#ifdef USE_PPOLL
+  pollfd pollfds[] = {{rfd_, POLLIN, 0}};
+  int ret = poll(pollfds, 1, 0);
+#else
+  fd_set set;
+  struct timeval timeout = { 0, 0 };
+  FD_ZERO(&set);
+  FD_SET(rfd_, &set);
+  int ret = select(rfd_ + 1, &set, NULL, NULL, &timeout);
+#endif
+  if (ret > 0) {
+    // Handle potential race condition:
+    //  - the above check succeeded, i.e. read() should not block
+    //  - the character disappears before we call read()
+    //
+    // Create a duplicate of rfd_. The duplicate file descriptor dup_rfd_
+    // can safely be closed by signal handlers without affecting rfd_.
+    dup_rfd_ = dup(rfd_);
+
+    if (dup_rfd_ != -1) {
+      struct sigaction act, old_act;
+      int ret = 0;
+      char buf;
+
+      // Temporarily replace SIGCHLD handler with our own
+      memset(&act, 0, sizeof(act));
+      act.sa_handler = CloseDupRfd;
+      if (sigaction(SIGCHLD, &act, &old_act) == 0) {
```

Comment:
Why?... Unlike `make`, `ninja` doesn't monitor the child as a child process via `SIGCHLD`, but merely tracks the pipe opened to it. Meaning it's not even subject to the same race conditions which `make` had to handle.

Reply:
Ah... this code used to provide non-blocking reads in a legacy portable way, not recognizing `O_NONBLOCK` as a legit way of doing this.

The `SIGCHLD` part is pretty much entirely irrelevant though - it's merely an optimization.